src/core/compact_object.cc

```diff
@@ -680,32 +691,49 @@ std::optional<int64_t> CompactObj::TryGetInt() const {
 
 auto CompactObj::GetJson() const -> JsonType* {
   if (ObjType() == OBJ_JSON) {
-    DCHECK_EQ(u_.json_obj.encoding, kEncodingJsonCons);
-    return u_.json_obj.json_ptr;
+    DCHECK_EQ(JsonEnconding(), kEncodingJsonCons);
+    return u_.json_obj.cons.json_ptr;
   }
   return nullptr;
 }
 
 void CompactObj::SetJson(JsonType&& j) {
-  if (taglen_ == JSON_TAG && u_.json_obj.encoding == kEncodingJsonCons) {
-    // already json
-    DCHECK(u_.json_obj.json_ptr != nullptr);  // must be allocated
-    u_.json_obj.json_len = j.size();
-    u_.json_obj.json_ptr->swap(j);
-  } else {
-    SetMeta(JSON_TAG);
-    u_.json_obj.json_len = j.size();
-    u_.json_obj.json_ptr = AllocateMR<JsonType>(std::move(j));
-    u_.json_obj.encoding = kEncodingJsonCons;
+  if (taglen_ == JSON_TAG && JsonEnconding() == kEncodingJsonCons) {
+    DCHECK(u_.json_obj.cons.json_ptr != nullptr);  // must be allocated
+    u_.json_obj.cons.json_ptr->swap(j);
+    // We do not set bytes_used as this is needed. Consider the two following cases:
+    // 1. old json contains 50 bytes. The delta for new one is 50, so the total bytes
+    // the new json occupies is 100.
+    // 2. old json contains 100 bytes. The delta for new one is -50, so the total bytes
+    // the new json occupies is 50.
+    // Both of the cases are covered in SetJsonSize and JsonMemTracker. See below.
+    return;
+  }
+
+  SetMeta(JSON_TAG);
+  u_.json_obj.cons.json_ptr = AllocateMR<JsonType>(std::move(j));
+  u_.json_obj.cons.bytes_used = 0;
+}
+
+void CompactObj::SetJsonSize(bool net_positive, size_t size) {
+  if (taglen_ == JSON_TAG && JsonEnconding() == kEncodingJsonCons) {
+    // JSON.SET or if mem hasn't changed from a JSON op then we just update.
+    if (u_.json_obj.cons.bytes_used == 0) {
+      u_.json_obj.cons.bytes_used = size;
+    } else if (net_positive) {
+      u_.json_obj.cons.bytes_used += size;
+    } else {
+      DCHECK(u_.json_obj.cons.bytes_used >= size);
+      u_.json_obj.cons.bytes_used -= size;
+    }
```

Comment:
Why not accept `size` (call it `delta`) as some signed type (`int64_t` or whatever), and then just do `u_.json_obj.cons.bytes_used += delta;`?
Why all the `if`s and extra parameter?

Reply:
Also specifically this first `if` is unneeded right?

Reply:
yes I agree we don't need the first case. Furthermore, I was afraid of overflowing/underflowing but we do the exact same thing as you suggest in `AccountObjectMemory` so I guess it should not be an issue. Delta sounds simpler also. I will apply this.